ethereum/listener.go

```diff
@@ -10,187 +10,360 @@ import (
+
+	// get history from (start block - lookback) up until latest block
+	latestBlock := e.LatestBlock()
+	start := latestBlock
+	if e.startBlock != 0 {
+		start = e.startBlock
+	}
+	startLookback := start - e.lookbackPeriod
+	logger.Info(fmt.Sprintf("Getting history from %d: starting at: %d looking back %d blocks", startLookback, start, e.lookbackPeriod))
+	e.getAndConsumeHistory(ctx, logger, startLookback, latestBlock)
+
+	logger.Info("Finished getting history")
 
-		e.startBlock = header.Number.Uint64()
+	if flushInterval > 0 {
+		go e.flushMechanism(ctx, logger, sig)
 	}
 
+	// listen for errors in the main websocket stream
+	// if error occurs, trigger sig.Ready
+	// This will cancel `consumeStream` and `flushMechanism` routines
+	select {
+	case <-ctx.Done():
+		return
+	case err := <-sub.Err():
+		logger.Error("Websocket disconnected. Reconnecting...", "err", err)
+		close(sig.Ready)
+
+		// restart
+		e.startBlock = e.lastFlushedBlock
+		time.Sleep(10 * time.Millisecond)
+		e.startListenerRoutines(ctx, logger)
+		return
+	}
+
+}
+
+func (e *Ethereum) startMainStream(
+	ctx context.Context,
+	logger log.Logger,
+) (stream <-chan ethtypes.Log, sub ethereum.Subscription, history []ethtypes.Log) {
+
+	var err error
+
+	etherReader := etherstream.Reader{Backend: e.wsClient}
+
+	latestBlock := e.LatestBlock()
+
+	// start initial stream (start-block and lookback period handled separately)
+	logger.Info("Starting Ethereum listener")
+
 	query := ethereum.FilterQuery{
 		Addresses: []common.Address{messageTransmitterAddress},
 		Topics:    [][]common.Hash{{messageSent.ID}},
-		FromBlock: big.NewInt(int64(e.startBlock - e.lookbackPeriod)),
+		FromBlock: big.NewInt(int64(latestBlock)),
 	}
 
-	logger.Info(fmt.Sprintf(
-		"Starting Ethereum listener at block %d looking back %d blocks",
-		e.startBlock,
-		e.lookbackPeriod))
+	queryAttempt := 1
+	for {
+		// websockets do not query history
+		// https://github.com/ethereum/go-ethereum/issues/15063
+		stream, sub, history, err = etherReader.QueryWithHistory(ctx, &query)
+		if err != nil {
+			logger.Error("Unable to subscribe to logs", "attempt", queryAttempt, "err", err)
+			queryAttempt++
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		break
+	}
 
-	// websockets do not query history
-	// https://github.com/ethereum/go-ethereum/issues/15063
-	stream, sub, history, err := etherReader.QueryWithHistory(ctx, &query)
-	if err != nil {
-		logger.Error("unable to subscribe to logs", "err", err)
-		os.Exit(1)
+	return stream, sub, history
+}
+
+func (e *Ethereum) getAndConsumeHistory(
+	ctx context.Context,
+	logger log.Logger,
+	start, end uint64) {
+
+	var toUnSub ethereum.Subscription
+	var history []ethtypes.Log
+	var err error
+
+	// handle historical queries in chunks (some websockets only allow small history queries)
+	chunkSize := uint64(100)
+	chunk := 1
+	totalChunksNeeded := (end - start) / chunkSize
+	if (end-start)%chunkSize > 0 || totalChunksNeeded == 0 {
+		totalChunksNeeded++
```

Comment:
Ensure that start > end errors? else you'll get `Total chunks needed:  184467440737095512`

Also if you want to simplify logic this helps remove the if statement
```go
const chunkSize = uint64(100)
 totalChunksNeeded := (end - start + chunkSize - 1) / chunkSize
```